### PR TITLE
Fix param passthrough

### PR DESCRIPTION
**Pull request by ziofil**
**Context:**
Circuits don't allow for correlated parameters

**Description of the Change:**
Objects accept parameters other than simply an initial value. This allows for circuits like e.g. `Sgate(r=BS.theta)`.

**Benefits:**
We can make circuits that live on submanifolds and we can correlate gates with each other.

**Possible Drawbacks:**
We should explain somewhere that if a parameter is a function of trainable variables, the function has to be computed inside the cost function (e.g. `BSgate(theta = f(vars))` has to run inside the cost function, otherwise `f(vars)` is not traced for the gradients).

**Related GitHub Issues:**
None

### CODE REVIEW

File: tests/test_utils/test_opt.py [hunk 1]
```diff
@@ -230,6 +235,29 @@ def cost_fn():
     assert np.allclose(np.sinh(circ.trainable_parameters["euclidean"][2]) ** 2, 1, atol=1e-2)
 
 
+def test_parameter_passthrough():
+    """Same as the test above, but with param passthrough"""
+    tf.random.set_seed(137)
+    r = np.arcsinh(1.0)
+    par = Parametrized(
+        r=math.new_variable(r, (0.0, None), "r"),
+        phi=math.new_variable(np.random.normal(), (None, None), "phi"),
```
Comment (sduquemesa) [inline tests/test_utils/test_opt.py:244]: Should this variable be parametrized with hypothesis instead of defined here?

**Review by sduquemesa — APPROVED** [review]
Great @ziofil! I just suggest to extend a bit the changelog to better describe the change.

On a more general point, I wonder if parameters should be better represented by classes themselves, in that way they can self-contain all the relevant information and avoid having to include conditional logic when dealing with them — a parameter object would know if it is trainable or constant, if already defined as a tensor in the backend or not, etc. 🤔

File: tests/test_utils/test_opt.py (hunk 1 above)
Comment (ziofil) [reply tests/test_utils/test_opt.py:244]: > Great @ziofil! I just suggest to extend a bit the changelog to better describe the change.
> 
> On a more general point, I wonder if parameters should be better represented by classes themselves, in that way they can self-contain all the relevant information and avoid having to include conditional logic when dealing with them — a parameter object would know if it is trainable or constant, if already defined as a tensor in the backend or not, etc. 🤔

Yeah that's a good idea. I've thought about implementing a Parameter class when I wanted to track the history of each parameter and plot it, but haven't found time to work on it yet : /

Comment (ziofil) [reply tests/test_utils/test_opt.py:244]: Thanks for the change in the changelog, it looks great!

**Pull request merged**